prov/verbs/src/windows/fi_verbs_nd.h

```diff
@@ -0,0 +1,259 @@
+	struct modules_t {
+		struct module_t *modules;
+		size_t count;
+	} providers;
+
+	struct class_factory_t {
+		struct factory_t *factory;
+		size_t count;
+	} class_factories;
+
+	struct adapters_t {
+		struct adapter_t *adapter;
+		struct adapter_t **adapter_list;
+		struct ibv_device **device_list;
+		struct ibv_context **context_list;
+		size_t count;
+	} adapters;
+};
+extern struct vrb_nd_infra_t vrb_nd_infra;
+
+inline struct adapter_t *get_adapter_by_context(struct ibv_context *context)
+{
+	for (int i = 0; i < vrb_nd_infra.adapters.count; ++i) {
+		if (vrb_nd_infra.adapters.context_list[i] == context) {
+			return vrb_nd_infra.adapters.adapter_list[i];
+		}
+	}
+	return NULL;
+}
+
+inline struct ibv_context *get_context_by_device(struct ibv_device *device)
+{
+	for (int i = 0; i < vrb_nd_infra.adapters.count; ++i) {
+		if (vrb_nd_infra.adapters.device_list[i] == device) {
+			return vrb_nd_infra.adapters.context_list[i];
+		}
+	}
+	return NULL;
+}
+
+// Wrapper structs
+struct rdma_cm_id_vrb_nd {
+	struct rdma_cm_id id;
+	HANDLE ov_file;
+	IND2Connector *connector;
+	IND2Listener *listener;
+	ofi_atomic32_t cm_events_pending;
+};
+
+struct rdma_event_channel_vrb_nd {
+	struct rdma_event_channel channel;
+	struct dlistfd_head q;
+};
+
+struct rdma_cm_event_vrb_nd {
+	struct rdma_cm_event event;
+	struct dlist_entry entry;
+};
+
+struct ibv_cq_vrb_nd {
+	struct ibv_cq cq;
+	HANDLE ov_file;
+	IND2CompletionQueue *nd2cq;
+};
+
+struct ibv_mr_vrb_nd {
+	struct ibv_mr mr;
+	HANDLE ov_file;
+	IND2MemoryRegion *region;
+};
+
+struct ibv_qp_vrb_nd {
+	struct ibv_qp qp;
+	IND2QueuePair *nd2qp;
+	uint32_t max_inline_data;
+};
+
+struct ibv_comp_channel_vrb_nd {
+	struct ibv_comp_channel channel;
+	uint32_t num_pending;
+	ofi_mutex_t q_lock;
+	struct dlistfd_head q;
+};
+
+struct ibv_comp_event_vrb_nd {
+	struct ibv_cq *cq;
+	void *cq_ctx;
+	struct dlist_entry entry;
+};
+
+typedef void (*vrb_nd_event_t)(struct vrb_nd_event_base *base, DWORD bytes);
+typedef void (*vrb_nd_error_t)(struct vrb_nd_event_base *base, DWORD bytes,
+			       DWORD error);
+
+typedef struct vrb_nd_event_base {
+	OVERLAPPED ov;
+	vrb_nd_event_t event_cb;
+	vrb_nd_error_t error_cb;
+} vrb_nd_event_base;
+
+void CALLBACK vrb_nd_io_cb(DWORD error, DWORD bytes, LPOVERLAPPED ov);
+
+// Used in conjunction with IND2Listener::GetConnectionRequest()
+typedef struct vrb_nd_cm_listen_event {
+	vrb_nd_event_base base;
+	IND2Connector *connector;
+	struct rdma_cm_id *listen_id;
+} vrb_nd_cm_listen_event;
+
+void vrb_nd_cm_connreq_event(vrb_nd_event_base *base, DWORD bytes);
+void vrb_nd_cm_connreq_error(vrb_nd_event_base *base, DWORD bytes, DWORD error);
+
+// Used in conjunction with
+//   IND2Connector::Connect()
+//   IND2Connector::Accept()
+//   IND2Connector::CompleteConnect()
+//   IND2Connector::Disconnect()
+//   IND2Connector::NotifyDisconnect()
+typedef enum vrb_nd_cm_state {
+	CONNECT,
+	ACCEPT,
+	COMPLETE,
+	DISCONNECT,
+	DISCONNECTED
+} vrb_nd_cm_state;
```

Comment:
No typedef and please add a prefix to the values.  There should be some IDLE, UNCONNECTED, or UNKNOWN state as the first value, since a newly allocated structure that uses this enum will not start as connected.